lib/bridge/IrcHandler.js

```diff
@@ -176,6 +176,66 @@ IrcHandler.prototype.onPrivateMessage = Promise.coroutine(function*(req, server,
     yield this.ircBridge.sendMatrixAction(pmRoom, virtualMatrixUser, mxAction, req);
 });
 
+
+/**
+ * Called when the AS receives an IRC topic event.
+ * @param {IrcServer} server : The sending IRC server.
+ * @param {IrcUser} fromUser : The sender.
+ * @param {string} channel : The target channel.
+ * @param {Object} action : The IRC action performed.
+ * @return {Promise} which is resolved/rejected when the request finishes.
+ */
+IrcHandler.prototype.onTopic = Promise.coroutine(function*(req, server, fromUser,
+                                                    channel, action) {
+    req.log.info("onTopic: %s from=%s to=%s action=%s",
+        server.domain, fromUser, channel, JSON.stringify(action).substring(0, 80)
+    );
+
+    // Only bridge topics for rooms created by the bridge, via !join or an alias
+    let origins = ["join", "alias"];
+    let mappings = yield this.ircBridge.getStore().getMappingsForChannelByOrigin(
+        server, channel, origins
+    );
+    if (mappings.length === 0) {
+        req.log.info(
+            "No mapped matrix rooms for IRC channel %s with origin = [%s]",
+            channel,
+            origins
+        );
+        return Promise.resolve();
+    }
+    let promises = mappings.map((entry) => {
+        if (entry.matrix.topic === action.text) {
+            req.log.info(
+                "Topic of %s already set to '%s'",
+                entry.matrix.getId(),
+                action.text
+            );
+            return;
+        }
+        entry.matrix.topic = action.text;
+        this.ircBridge.getStore().upsertRoomStoreEntry(entry);
```

Comment:
This isn't safe to be done here. There's a lot of race conditions, but unfortunately this one is a big one.

Assuming the IRC Event Broker is doing its job, we *shouldn't* get `onTopic` called multiple times for the *same* topic change (1 per connection), so `onTopic` will only be called once and only once when the topic changes. So far so good.

When it does get called, we'll try to the following *concurrently*:
 - Update the DB
 - Set the topic via HTTP

What if the HTTP request fails? The topic will be in the DB but will not be set on the room! When should we be updating the DB?

This gets exacerbated if a real IRC user changes the topic very quickly (e.g. `/topic 1`, `/topic 2`, `/topic 3`). Is there any guarantee that "3" will be the final topic set on Matrix?